packages/expo-background-task/android/src/main/java/expo/modules/backgroundtask/BackgroundTaskConsumer.kt

```diff
@@ -0,0 +1,116 @@
+import expo.modules.interfaces.taskManager.TaskInterface
+import expo.modules.interfaces.taskManager.TaskManagerUtilsInterface
+import expo.modules.interfaces.taskManager.TaskServiceProviderHelper
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.cancel
+import kotlinx.coroutines.launch
+
+const val BACKGROUND_TASK_TYPE = "expo-background-task"
+const val DEFAULT_INTERVAL_MINUTES = 60L * 24L // Once every day
+private val TAG = BackgroundTaskConsumer::class.java.simpleName
+
+class BackgroundTaskConsumer(context: Context?, taskManagerUtils: TaskManagerUtilsInterface?) :
+  TaskConsumer(context, taskManagerUtils), TaskConsumerInterface, LifecycleEventListener {
+
+  private var task: TaskInterface? = null
+  private var isBackgrounded: Boolean = true
+  private val taskCoroutineScope = CoroutineScope(Dispatchers.Default)
+
+  override fun taskType(): String {
+    return BACKGROUND_TASK_TYPE
+  }
+
+  /**
+   * Exposing the execute task function so that the BackgroundTaskWork (WorkManager Work unit)
+   * can execute the inner task. A task is only executed if the app is in the background.
+   * This only applies when the app is in release mode.
+   */
+  fun executeTask(callback: TaskExecutionCallback) {
+    if (isBackgrounded || ReactBuildConfig.DEBUG) {
+      Log.i(TAG, "Executing task '${task?.name}'")
+      taskManagerUtils.executeTask(task, null, callback)
+    }
+  }
+
+  override fun didRegister(task: TaskInterface) {
+    this.task = task
+
+    Log.i(TAG, "didRegister: ${task.name}")
+
+    // Make sure to schedule and set up the worker if it is not running
+    taskCoroutineScope.launch {
+      val context = context
+      if (!BackgroundTaskScheduler.isWorkerRunning(context)) {
+        // Get interval for the task
+        val intervalMinutes = getIntervalMinutes()
+        // Start worker
+        Log.i(TAG, "didRegister: worker not running - starting worker.")
+        BackgroundTaskScheduler.startWorker(context, task.appScopeKey, intervalMinutes)
+      } else {
+        Log.i(TAG, "didRegister: worker already running.")
+      }
+    }
+  }
+
+  override fun didUnregister() {
+    Log.i(TAG, "didUnregister: ${task?.name}")
+
+    taskCoroutineScope.launch {
+      if (!BackgroundTaskScheduler.isWorkerRunning(context)) {
+        return@launch
+      }
+
+      // Check if the task manager has more tasks like this
+      val taskService = TaskServiceProviderHelper.getTaskServiceImpl(context.applicationContext)
+        ?: throw MissingTaskServiceException()
+
+      // Get tasks for our appScope
+      val appScopeKey = task?.appScopeKey ?: throw MissingAppScopeKey()
+      task = null
+
+      // Check if we have a task that
+      val tasks = taskService.getTasksForAppScopeKey(appScopeKey)
+      val ourTasks = tasks.filter { it.getString("taskType") == BACKGROUND_TASK_TYPE }
+      if (ourTasks.isEmpty()) {
```

Comment:
note that `Dispatchers.Default` is a [thread pool](https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines/-dispatchers/-default.html). if you intend to execute this code in concurrent mode, maybe you should add a lock here. otherwise, maybe try to use other coroutine with single thread.